carapace-server/src/main/java/org/carapaceproxy/server/Listeners.java

```diff
@@ -154,17 +169,26 @@ private SslContext bootSslContext(NetworkListenerConfiguration listener, SSLCert
                 LOG.log(Level.INFO, "required sslCiphers " + sslCiphers);
                 ciphers = Arrays.asList(sslCiphers.split(","));
             }
-            return SslContextBuilder
-                            .forServer(keyFactory)
-                            .enableOcsp(listener.isOcsp() && OpenSsl.isOcspSupported())
-                            .trustManager(trustManagerFactory)
-                            .sslProvider(SslProvider.OPENSSL)
-                            .protocols(listener.getSslProtocols())
-                            .ciphers(ciphers).build();
+            SslContext sslContext = SslContextBuilder
+                    .forServer(keyFactory)
+                    .enableOcsp(listener.isOcsp() && OpenSsl.isOcspSupported())
+                    .trustManager(trustManagerFactory)
+                    .sslProvider(SslProvider.OPENSSL)
+                    .protocols(listener.getSslProtocols())
+                    .ciphers(ciphers).build();
+
+            if (listener.isOcsp() && OpenSsl.isOcspSupported()) {
+                String certKey = certificate.getId();
+                parent.getOcspStaplingManager().addCertificateForStapling(certKey, chain);
```

Comment:
This method is called when a client connects to the proxy.
We should issue the OSCP call in a separate operation:
- at boot -> issue all of the OSCP requests (in background)
- periodically -> issue all of expired OSCP requests

With the current approach we can have race conditions with concurrent clients asking for a particular certificate.

I feel it is not so important to always put the ocspresponse, for instance when the system is cold (right after bootstrap)  it is acceptable that some client does not receive the OSCP response

Reply:
@eolivelli yes, it's acceptable to not always send OCSP data (in this case the clients will check directly with the CAs to see if the cert is still valid). OCSP stapling is "only" a performance improvement for the client, reliability of the system is far more important (and I've seen instances of Nginx and other webserver not honoring OCSP stapling)